tests/io/xml/sdmx21/writer/test_structures_writing.py

```diff
@@ -167,3 +168,67 @@ def test_writer_no_header():
     assert "<mes:Test>true</mes:Test>" in result
     assert "<mes:Prepared>" in result
     assert '<mes:Sender id="ZZZ"/>' in result
+
+
+def test_writer_datastructure(complete_header):
+    datastructure = DataStructureDefinition(
+        agency=Agency(
+            id="BIS",
+            name="Bank for International Settlements",
+            urn="urn:sdmx:org.sdmx.infomodel.datastructure."
+            "DataStructure=BIS:BIS_DER(1.0)",
+        ),
+        annotations=(),
+        id="BIS_DER",
+        description="BIS derivates statistics",
+        is_external_reference=None,
+        is_final=None,
+        name="BIS derivates statistics",
+        service_url=None,
+        structure_url=None,
+        uri="http://www.bis.org/statistics/derivatives.html",
+        urn="urn:sdmx:org.sdmx.infomodel.datastructure."
+        "DataStructure=BIS:BIS_DER(1.0)",
+        valid_from=datetime.strptime("2021-01-01", "%Y-%m-%d"),
+        valid_to=datetime.strptime("2021-12-31", "%Y-%m-%d"),
+        version="1.0",
+    )
+
+    result = writer(
+        {"DataStructures": {"FREQ": datastructure}},
+        MessageType.Structure,
+        header=complete_header,
+        prettyprint=True,
+    )
+
+    assert "DataStructure=BIS:BIS_DER(1.0)" in result
+
+
+def test_writer_dataflow(complete_header):
+    dataflow = Dataflow(
+        agency="BIS",
+        annotations=(),
+        id="WEBSTATS_DER_DATAFLOW",
+        description="OTC derivatives and FX spot - turnover",
+        is_external_reference=None,
+        is_final=None,
+        name="OTC derivatives turnover",
+        service_url=None,
+        structure="Dataflow=BIS:WEBSTATS_DER_DATAFLOW(1.0)",
+        structure_url=None,
+        uri=None,
+        urn="urn:sdmx:org.sdmx.infomodel.datastructure."
+        "Dataflow=BIS:WEBSTATS_DER_DATAFLOW(1.0)",
+        valid_from=datetime.strptime("2021-01-01", "%Y-%m-%d"),
+        valid_to=datetime.strptime("2021-12-31", "%Y-%m-%d"),
+        version="1.0",
+    )
+
+    result = writer(
+        {"Dataflows": {"FREQ": dataflow}},
+        MessageType.Structure,
+        header=complete_header,
+        prettyprint=True,
+    )
+
+    assert "Dataflow=BIS:WEBSTATS_DER_DATAFLOW(1.0)" in result
```

Comment:
Same as above, we must compare to a reference

Reply:
Added read_write and write_read tests that compares output to all input content.